runtime/executor/src/test/java/org/apache/nemo/runtime/executor/task/TaskExecutorTest.java

```diff
@@ -601,6 +600,91 @@ public void close() {
     }
   }
 
+  /**
+   * Source vertex for unbounded source test.
+   */
+  private final class TestUnboundedSourceVertex extends SourceVertex {
+
+    @Override
+    public boolean isBounded() {
+      return false;
+    }
+
+    @Override
+    public List<Readable> getReadables(int desiredNumOfSplits) throws Exception {
+      return null;
+    }
+
+    @Override
+    public void clearInternalStates() {
+
+    }
+
+    @Override
+    public IRVertex getClone() {
+      return null;
+    }
+  }
+
+  private final class TestUnboundedSourceReadable implements Readable {
+    int pointer = 0;
+    final int middle = elements.size() / 2;
+    final int end = elements.size();
+    boolean watermarkEmitted = false;
+    final List<Long> watermarks;
+    int numEmittedWatermarks = 0;
+
+    public TestUnboundedSourceReadable(final List<Long> watermarks) {
+      this.watermarks = watermarks;
+    }
+
+    @Override
+    public void prepare() {
+
+    }
+
+    // This emulates unbounded source that throws NoSuchElementException
+    // It reads current data until middle point and  throws NoSuchElementException at the middle point.
+    // It resumes the data reading after emitting a watermark, and finishes at the end of the data.
+    @Override
+    public Object readCurrent() throws NoSuchElementException {
+      if (pointer == middle && !watermarkEmitted) {
```

Comment:
Can you explain a little bit more in detail (in the comment) why `! watermarkEmitted` is needed here?
`readCurrent` and watermark seem independent operations to me.

Reply:
The watermark is triggered by TaskExecutor internally every a certain period so I want to block reading a value until the watermark is emitted.